### DISCUSSION TITLE
Issues with DataPortal.UpdateChild method in case of BusinessListBase child class

### DISCUSSION THREAD

**Discussion by KetanBhojak** [MarimerLLC/csla#2449] [Questions]
Hi Team,

We have been using CSLA in our Asp.Net Web application since 12-13 years. We are using CSLA 4.3.13.0 in our application since long. Now, recently we decided to upgrade the CSLA Version on our application. We initially used CSLA 5.4.0 and then eventually upgraded to CSLA 5.5.0 and 5.5.1 (which is just released last week). This upgrade activity is still in development phase and we are seeing what all changes we would need in our application to work with latest DLLs and we are also updating things at several places.

On this activity, we are now left with one type of issue with DataPortal.UpdateChild method. Please find below details about the issue:

CSLA version: 5.5.1.0
OS: Windows
Platform: ASP.Net application

We have following classes:

1. BusinessBase class HotelRoom.cs with HotelRoomID INT and HotelRoomName String properties
2. BusinessListBase class HotelRoomList.cs of the type HotelRoom to accommodate multiple HotelRooms
3. BusinessBase class Hotel.cs with Properties HotelID INT, HotelName String and HotelRoomList of BusinessListBase type for multiple HotelRooms with single Hotel
4. Hotel.Server.cs partial class with all the implementation of Inserting, Deleting and Updating Hotels
5. HotelRoom.Server.cs partial class with all implementation of Inserting, Deleting and Updating HotelRooms from a particular Hotel. As the HotelRoom is child of Hotel class, we have Child_Insert, Child_Update and Child_DeleteSelf implementation in HotelRoom.server.cs class with calls of DB for adding/updating and deleting HotelRooms.

From UI, we have pages to add/update hotels and also pages for insert/update/delete hotelrooms as child of Hotel. So, let's consider a scenario where we have a Hotel adding in the system with zero hotel rooms. Now, we move to the page to add HotelRooms under that Hotel. So, we have some code as below:

```c#
Hotel h = Hotel.GetHotel(hotelID);
HotelRoom hr = HotelRoom.NewHotelRoom() (this is constructor with DataPortal.CreateChild)
hr.HotelRoomName = "abc";
h.HotelRoomList.Add(hr);
h.Save();
```

So, above h.Save() line goes to 'DataPortal_Update' of Hotel.cs class; where we have also called below line to update the child list of HotelRoom:

```c#
DataPortal.UpdateChild(HotelRoomList, this);
```

Now, the above line is supposed to go to a proper method of HotelRoom.Server.cs class (either Child_Insert, Child_Update or Child_DeleteSelf) based on different flags of the object (IsNew, IsDirty, IsDeleted etc). In our application, in case of Inserting HotelRoom, Updating or Deleting HotelRoom, we are maintaining the same pattern of getting Hotel object and then updating the HotelRoomList property of hotel object and then hitting "DataPortal.UpdateChild(HotelRoomList, this);" line in all cases.

Until we were using CSLA 4.3.13.0, we were successfully able to use above pattern and the flow was correctly going into Child_Insert, Child_Update or Child_DeleteSelf method of HotelRoom.server.cs class.
But since we are using CSLA 5.4.0 or 5.5.0 and 5.5.1, the above implementation is not working for us. In the above case of inserting HotelRoom, system throws below exception when we call "DataPortal.UpdateChild(HotelRoomList, this);" line to insert/update the HotelRoom child:

```text
**Exception**: {"ChildDataPortal.Update failed on the server"}	Csla.DataPortalException
**BusinessException**: Application.HotelRoom.[InsertChild](Hotel)
**InnerException.Message**: ChildDataPortal.Update failed on the server
**StackTrace**:    at Csla.Reflection.ServiceProviderMethodCaller.FindDataPortalMethod[T](Type targetType, Object[] criteria, Boolean throwOnError)
   at Csla.Reflection.LateBoundObject.<CallMethodTryAsyncDI>d__12 1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Csla.Server.DataPortalTarget.<UpdateChildAsync>d__20.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Csla.Server.ChildDataPortal.<Update>d__18.MoveNext()
   at Csla.Reflection.ServiceProviderMethodCaller.<CallMethodTryAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Csla.Reflection.LateBoundObject.<CallMethodTryAsyncDI>d__12 1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Csla.Server.DataPortalTarget.<UpdateChildAsync>d__20.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Csla.Server.ChildDataPortal.<Update>d__18.MoveNext()
   at Csla.Server.ChildDataPortal.Update(Object obj, Object[] parameters)
   at Csla.Reflection.ServiceProviderMethodCaller.<CallMethodTryAsync>d__11.MoveNext()
Inner Exception 1: CallMethodException: HotelRoomList.Child_Update method call failed
Inner Exception 2: TargetParameterCountException: Application.HotelRoom.[InsertChild](Hotel)
Inner Exception 3: TargetParameterCountException: Csla.BusinessBase 1[[Application.HotelRoom, Application, Version=6.2.0.0, Culture=neutral, PublicKeyToken=442253b88929624d]].[InsertChild](Hotel)
Inner Exception 4: TargetParameterCountException: Csla.Core.BusinessBase.[InsertChild](Hotel)
Inner Exception 5: TargetParameterCountException: Csla.Core.UndoableBase.[InsertChild](Hotel)
Inner Exception 6: TargetParameterCountException: Csla.Core.BindableBase.[InsertChild](Hotel)
Inner Exception 7: TargetParameterCountException: Csla.Core.MobileObject.[InsertChild](Hotel)
Inner Exception 8: TargetParameterCountException: System.Object.[InsertChild](Hotel)
```

Now, as the above implementation is not working anymore, we did some analysis and went through release notes of version between 4.3.13.0 to 5.5.1.0; but we couldn't find any exact point where above implementation is changed (though we could find out some changes related to UpdateChild method). The above implementation works just fine with CSLA 4.3.13.0.

While finding solution to this, we could observe that the things was working fine if we replace line from "DataPortal.UpdateChild(HotelRoomList, this);" to "DataPortal.UpdateChild(HotelRoomList);". But it then breaks update case of HotelRoom object, and hence we have to keep both version of above line, separate for Insert and Update operations. This is not feasible for us to do as we have above type of implementation at hundreds of places in our application.

Can someone guide us on this and let us know what we can do for this and why this line has stopped working with CSLA 5.5.1.0. Also, if there are any plan to address this in next release of CSLA.

**Comment by rockfordlhotka** [comment]
From CSLA 4.x to 5.x the way the data portal finds and invokes methods was changed _at lot_ to implement the ability to `Inject` parameters to data portal operation methods.

In 5.5.0 and 5.5.1 there was a change, but only to the DeleteChildSelf operation: #2396

So I'm not sure why this is failing for you. We need to create a unit test to reproduce the failure. Ideally a new test or tests here

https://github.com/MarimerLLC/csla/blob/v5.x/Source/csla.netcore.test/DataPortal/ChildDataPortalTests.cs

**Comment by KetanBhojak** [comment]
Thank you Rockford for your quick reply.

Just to clarify one thing here, we are having Asp.Net Web Form Application; not the Asp.Net Core application. And hence, we have used NuGet packages from "https://www.nuget.org/packages/CSLA-ASP.NET/5.0.0-R19040402"

Do let me know if you need any more details or information from me on this to replicate the issue.

**Comment by rockfordlhotka** [comment]
So you are using the prerelease of 5.0.0, not the prerelease of 5.5.1 or the final release of 5.5.0?

In any case, the data portal method invocation tests are in that file I referenced - the implementation is the same for all platforms.

**Comment by KetanBhojak** [comment]
Hello Rockford,

There was a typo in the URL of NuGet I posted in my last comment. We are actually using Pre release of 5.5.1 version (but the issue I am talking about, is same in 5.5.0 and 5.4.0 as well). We got NuGet from below URL "https://www.nuget.org/packages/Csla.AspNet/5.5.1-R21080301".

Apart from this, we have created a small Asp.Net application to replicate the exact issue. I have attached zip file of the entire solution. In that, the Hotel.Server.cs class breaks on UpdateChild method for child list. We have kept the logic on Page_Load of default.aspx page only, so the logic will execute on the application startup only.

I have removed 'packages' folder due to size constraint, but as I said, we have used 5.5.1 Pre release version of CSLA. I am attaching the solution zip file here for the team to replicate the issue and provide us further suggestions.

[CSLAUpgrade_TestApp.zip](https://github.com/MarimerLLC/csla/files/6976679/CSLAUpgrade_TestApp.zip)

**Comment by KetanBhojak** [comment]
Hi Team,

Any updates on progress on this one?

**Comment by rockfordlhotka** [comment]
My "day job" has been very demanding of late, hopefully I'll get some quality CSLA time yet this week.

**Comment by rockfordlhotka** [comment]
When you call `DataPortal.UpdateChild(HotelRoomList, this);` in the `Hotel` class, the data portal will call the `Child_Update` method of the `HotelRoomList` object, passing it one parameter (`this`).

In the `HotelRoomList` class you can override that method to see what is passed in as a parameter:

```c#
        protected override void Child_Update(params Object[] parameters)
        {
            base.Child_Update(parameters);
        }
```

Notice that this method (by default) then calls the base class method to update all child in the collection, passing that same list of parameters. As a result, the _expectation_ is that each item in the collection will have a child insert/update/delete method that accepts that same list of parameters.

The assumption being that if you passed a reference to `Hotel` in the original call, that you meant for that reference to be available to all child objects being saved.

```c#
        [InsertChild]
        private void Child_Insert(Hotel h)
        {
            int a = 1;
        }

        [UpdateChild]
        private void Child_Update(Hotel h)
        {
            int a = 1;
        }

        [DeleteSelfChild]
        private void Child_DeleteSelf(Hotel h)
        {
            int a = 1;
        }
```

**Comment by dchuen** [comment]
I experience a similar problem while upgrading from CSLA 4.11.2 to 5.5.1.  

In 4.11.2, the below function (in an editable child) without any parameter would work perfectly 

        private void Child_Update()

In 5.5.1, however, I got the exact same error 

    Inner Exception 8:
    TargetParameterCountException: System.Object.[UpdateChild](ParentEdit)

If I add the following function to the child, it will work fine.

        private void Child_Update(object parent)
        {
            Child_Update();
        }

It makes sense to pass the parent object to Child_Update, I can see it can be useful sometime.  Is it possible to make this parameter optional to make it backward compatible?

**Reply by rockfordlhotka** [reply]
I see that we did decide to implement a fix (hack?) to address this backward compat issue, but only for the DeleteSelf methods.

https://github.com/MarimerLLC/csla/issues/2396

The same approach could be used for the ChildUpdate operation as well.

**Reply by rockfordlhotka** [reply]
I've recorded this in the backlog, so if anyone has some time to make the change and add a unit test that would be fantastic.

https://github.com/MarimerLLC/csla/issues/2559
